http/client/src/main/java/org/wildfly/security/http/client/ElytronHttpClient.java

```diff
@@ -0,0 +1,128 @@
+import java.util.Map;
+
+/**
+ * Elytron client for HTTP authentication
+ *
+ * @author <a href="mailto:kekumar@redhat.com">Keshav Kumar</a>
+ */
+public class ElytronHttpClient {
+
+    private HttpClient httpClient;
+    private static ElytronHttpClientCredentialUtils elytronHttpClientCredentialProvider = new ElytronHttpClientCredentialUtils();
+
+    public ElytronHttpClient() {
+        this.httpClient = HttpClient.newHttpClient();
+    }
+
+    private HttpResponse getResponse(HttpRequest request) throws IOException, InterruptedException{
+        return httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+    }
+
+    private void addSSLContextToHttpCient(URI uri){
+        SSLContext sslContext = elytronHttpClientCredentialProvider.getSSLContext(uri);
+        httpClient = HttpClient.newBuilder().sslContext(sslContext).build();
+    }
+    private HttpRequest evaluateMechanism(URI uri) {
+        HttpRequest request = HttpRequest
+                .newBuilder()
+                .uri(uri)
+                .build();
+
+        return request;
+    }
+
+    /**
+     * Used to connect to the secured uri and return the response based on that.
+     */
+    public HttpResponse connect(String uri) throws IOException, InterruptedException, URISyntaxException {
+
+        URI uriPath = new URI(uri);
+
+        String protocol = new URL(uri).getProtocol();
+
+        if(protocol.equalsIgnoreCase("https")){
+            addSSLContextToHttpCient(uriPath);
+        }
+
+        HttpRequest request = evaluateMechanism(uriPath);
+        HttpResponse response = getResponse(request);
+
+        if(response.statusCode() == OK){
+            return response;
+        }
+
+        String authHeader = getAuthHeader(response);
+
+        if(authHeader == null){
+            throw new ElytronHttpClientException(ElytronMessages.log.responseHeaderExtractionFailed());
+        }
+
+        String authType = authHeader.split(" ")[0].toLowerCase();
+
+        switch (authType){
+            case "basic" :
+                request = ElytronHttpClientBasicAuthMechanism.evaluateMechanism(uriPath);
+                break;
+            case "digest" :
+                request = ElytronHttpClientDigestAuthMechanism.evaluateMechanism(uriPath, authHeader);
+                break;
+            case "bearer" :
+                request = ElytronHttpClientBearerAuthMechanism.evaluateMechanism(uriPath);
+                break;
+            default:
+                request = evaluateMechanism(uriPath);
+        }
+
+        response = getResponse(request);
+        return response;
+    }
+
+    private String getAuthHeader(HttpResponse response){
+        String authHeader = null;
+        Map<String, List<String>> allHeaderValues = response.headers().map();
+        for(String headerKey : allHeaderValues.keySet()){
+            if(headerKey.toLowerCase().equals("www-authenticate")){
+                authHeader = allHeaderValues.get(headerKey).get(0);
```

Comment:
@keshav-725 Just a minor, the WWW-Authenticate header can contain multiple challenges https://datatracker.ietf.org/doc/html/rfc7235#section-4.1 , we should not ignore them